### PR TITLE
[compiler-rt][ASan] Remove alignment message in ASan error reporting

### PR DESCRIPTION
**Pull request by AdvenamTacet**
This commit removes unnecessary alignment check and message in ASan error reporting functions (like `ErrorBadParamsToAnnotateContiguousContainer::Print()`), as alignment is no longer required starting from LLVM 16.

Without that commit, this message can be observed only when arguments are truly incorrect and `beg` is unaligned. Just unaligned `beg` does not result in any message being printed.

Related commits:
- https://github.com/llvm/llvm-project/commit/dd1b7b797a116eed588fd752fbe61d34deeb24e4
- https://github.com/llvm/llvm-project/commit/1c5ad6d2c01294a0decde43a88e9c27d7437d157

### CODE REVIEW

**Comment by vitalybuka** [comment]
Would you mind to create a test to hit this check?

**Review by vitalybuka — CHANGES_REQUESTED** [review]
Test if possible
or explanation if it's hard

**Comment by AdvenamTacet** [comment]
This week I have no access to my computer, but I have reserved some time next week for open source related tasks and I will look at it.

It shouldn't be a problem to create an example, incorrect arguments (eg. end < beg) & unaligned beg should print it. 

But this message is simply obsolete, even if impossible to trigger. Unaligned buffers are supported.

**Comment by AdvenamTacet** [comment]
Example showing that `__sanitizer_annotate_contiguous_container` works with unaligned buffers: [godbolt](https://godbolt.org/#g:!((g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:15,positionColumn:1,positionLineNumber:1,selectionStartColumn:2,selectionStartLineNumber:15,startColumn:1,startLineNumber:1),source:'%23include+%3Csanitizer/asan_interface.h%3E%0A%23include+%3Ciostream%3E%0A%0Aconstexpr+size_t+N+%3D+64%3B%0Aconstexpr+size_t+off+%3D+1%3B%0A%0Aint+main()+%7B%0A%0A++++char+buffer%5BN+%2B+off%5D%3B%0A++++char+*beg+%3D+buffer+%2B+off%3B%0A++++char+*end+%3D+beg+%2B+N%3B%0A%0A++++__sanitizer_annotate_contiguous_container(beg,+end,+end,+beg)%3B%0A++++std::cout+%3C%3C+beg%5B0%5D+%3C%3C+std::endl%3B%0A%7D'),l:'5',n:'1',o:'C%2B%2B+source+%231',t:'0')),k:62.5,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+13.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+18.1.0+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:62.5,l:'3',n:'0',o:'',t:'0'),(g:!((h:compiler,i:(compiler:clang1810,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-fsanitize%3Daddress+-std%3Dlibc%2B%2B+-std%3Dc%2B%2B20',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+18.1.0+(Editor+%231)',t:'0')),header:(),k:37.5,l:'4',m:100,n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

```cpp
#include <sanitizer/asan_interface.h>
#include <iostream>

constexpr size_t N = 64;
constexpr size_t off = 1;

int main() {

    char buffer[N + off];
    char *beg = buffer + off;
    char *end = beg + N;

    __sanitizer_annotate_contiguous_container(beg, end, end, beg);
    std::cout << beg[0] << std::endl;
}
```

This example correctly detects container overflow (whole buffer is poisoned).
```python
=>0x7a1ccb000000: f1 f1 f1 f1[01]fc fc fc fc fc fc fc fc f3 f3 f
```

To get this message, it's enough to provide incorrect arguments to the function, for example providing `old_mid` outside of the container, like `__sanitizer_annotate_contiguous_container(beg, end, end+1, beg);`.
Then [we get](https://godbolt.org/z/4Yrsvd8es):
```cpp
==1==ERROR: AddressSanitizer: bad parameters to __sanitizer_annotate_contiguous_container:
      beg     : 0x74ab7b600021
      end     : 0x74ab7b600061
      old_mid : 0x74ab7b600062
      new_mid : 0x74ab7b600021
==1==ERROR: beg is not aligned by 8
```

Clearly `beg` not being aligned by 8 is not a problem here.

---

Those are only two lines with that error message.
```bash
$ grep -R -I -H "is not aligned by"
compiler-rt/lib/asan/asan_errors.cpp:    Report("ERROR: beg is not aligned by %zu\n", granularity);
compiler-rt/lib/asan/asan_errors.cpp:    Report("ERROR: storage_beg is not aligned by %zu\n", granularity);
llvm/test/tools/llvm-objcopy/ELF/binary-paddr.test:## PAddr is not aligned by sh_addralign(.data)
$
```

---

@vitalybuka do you think adding anything to that PR makes sense? If yes, what test do you have in mind?

**Comment by vitalybuka** [comment]
> ```c++
> #include <sanitizer/asan_interface.h>
> #include <iostream>
> 
> constexpr size_t N = 64;
> constexpr size_t off = 1;
> 
> int main() {
> 
>     char buffer[N + off];
>     char *beg = buffer + off;
>     char *end = beg + N;
> 
>     __sanitizer_annotate_contiguous_container(beg, end, end, beg);
> }
> ```

Just add above as a test,  it should exit with 0?

**Comment by AdvenamTacet** [comment]
Yes, it should exit with zero. I can add it, but we test it in contiguous_container.cpp test already.

---
Creating unaligned beg:

https://github.com/llvm/llvm-project/blob/055e4319112282354327af9908091fdb25149e9b/compiler-rt/test/asan/TestCases/contiguous_container.cpp#L123

Using unaligned beg:

https://github.com/llvm/llvm-project/blob/055e4319112282354327af9908091fdb25149e9b/compiler-rt/test/asan/TestCases/contiguous_container.cpp#L137-L138

---

Same for `__sanitizer_annotate_contiguous_container`:

https://github.com/llvm/llvm-project/blob/055e4319112282354327af9908091fdb25149e9b/compiler-rt/test/asan/TestCases/contiguous_container.cpp#L49-L59

**Comment by vitalybuka** [comment]
> Yes, it should exit with zero. I can add it, but we test it in contiguous_container.cpp test already.
> 
I am confused. Then why they are not triggering these messages without your patch?

**Comment by AdvenamTacet** [comment]
This patch here only removes printing the message.
Whenever incorrect arguments are provided `ErrorBadParamsToAnnotateContiguousContainer::Print()` is called, and all arguments are printed etc.
Additionally, inside that `Print`, if `beg` is unaligned, that message is printed.

However, because unaligned `beg` is correct, it does not trigger any message. The message is printed if another argument is incorrect and additionally beg is not aligned (and therefore is confusing).

**Comment by vitalybuka** [comment]
Idea is if someone undo your patch, it would be nice to have a test failure. As-is it's not going to happen.

**Review by vitalybuka — APPROVED** [review]
Never mind.
You don't have a test. That one godbolt does not trigger messages (on the HEAD).

**Comment by AdvenamTacet** [comment]
Now I am confused. On trunk we can see the message, but [at least one argument must be truly incorrect](https://godbolt.org/z/4ejcovnWE), same with older LLVM versions.

What we can do is add a check to [crash test](https://github.com/llvm/llvm-project/blob/055e4319112282354327af9908091fdb25149e9b/compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp).

```cpp
void BadBoundsUnaligned() {
  long t[100];
// CHECK-BAD-BOUNDS-UNALIGNED: ERROR: AddressSanitizer: bad parameters to __sanitizer_annotate_contiguous_container
// CHECK-BAD-BOUNDS-UNALIGNED: <rule telling that "beg is not aligned by" does not appear in the error message>
  __sanitizer_annotate_contiguous_container(&t[1], &t[0] + 100, &t[0] + 101,
                                            &t[0] + 50);
}
```

To confirm that this message does not contain "beg is not aligned by".

If you think it makes sense, I will add it.

**Comment by vitalybuka** [comment]
> Now I am confused. On trunk we can see the message, but [at least one argument must be truly incorrect](https://godbolt.org/z/4ejcovnWE), same with older LLVM versions.

Sorry, I clicked the first https://godbolt.org/ link which was messaging on 16 only, and forgot that it had instruction in the comment.

> 
> What we can do is add a check to [crash test](https://github.com/llvm/llvm-project/blob/055e4319112282354327af9908091fdb25149e9b/compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp).
> 
> ```c++
> void BadBoundsUnaligned() {
>   long t[100];
> // CHECK-BAD-BOUNDS-UNALIGNED: ERROR: AddressSanitizer: bad parameters to __sanitizer_annotate_contiguous_container
> // CHECK-BAD-BOUNDS-UNALIGNED: <rule telling that "beg is not aligned by" does not appear in the error message>
>   __sanitizer_annotate_contiguous_container(&t[1], &t[0] + 100, &t[0] + 101,
>                                             &t[0] + 50);
> }
> ```
> 
> To confirm that this message does not contain "beg is not aligned by".
> 
> If you think it makes sense, I will add it.

Yes, please.

**Comment by AdvenamTacet** [comment]
@vitalybuka do you know how correctly create "does not appear" rule? I added `-NOT-`, but it seems to not work.

File: compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp [hunk 1]
```diff
@@ -35,6 +36,13 @@ void BadBounds() {
                                             &t[0] + 50);
 }
 
+void UnalignedBadBounds() {
+  char t[100];
+// CHECK-NOT-UNALIGNED-BAD-BOUNDS: beg is not aligned by
```
Comment (Enna1) [inline compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp:41]: CHECK-NOT-UNALIGNED-BAD-BOUND-**NOT**

Comment (AdvenamTacet) [reply compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp:41]:
```suggestion
// CHECK-NOT-UNALIGNED-BAD-BOUNDS-NOT: beg is not aligned by
```

File: compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp [hunk 2]
```diff
@@ -35,6 +36,13 @@ void BadBounds() {
                                             &t[0] + 50);
 }
 
+void UnalignedBadBounds() {
+  char t[100];
+// CHECK-NOT-UNALIGNED-BAD-BOUNDS-NOT: beg is not aligned by
```
Comment (AdvenamTacet) [inline compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp:41]:
```suggestion
// CHECK-NOT-UNALIGNED-BAD-BOUNDS-NOT: beg is not aligned by
// CHECK-UNALIGNED-BAD-BOUNDS: ERROR: AddressSanitizer: bad parameters to __sanitizer_annotate_contiguous_container
```

File: compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp [hunk 3]
```diff
@@ -1,6 +1,7 @@
 // RUN: %clangxx_asan -O %s -o %t
 // RUN: not %run %t crash 2>&1 | FileCheck --check-prefix=CHECK-CRASH %s
 // RUN: not %run %t bad-bounds 2>&1 | FileCheck --check-prefix=CHECK-BAD-BOUNDS %s
+// RUN: not %run %t unaligned-bad-bounds 2>&1 | FileCheck --check-prefix=CHECK-NOT-UNALIGNED-BAD-BOUNDS %s
```
Comment (AdvenamTacet) [inline compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp:4]:
```suggestion
// RUN: not %run %t unaligned-bad-bounds 2>&1 | FileCheck --check-prefix=CHECK-NOT-UNALIGNED-BAD-BOUNDS %s
// RUN: not %run %t unaligned-bad-bounds 2>&1 | FileCheck --check-prefix=CHECK-UNALIGNED-BAD-BOUNDS %s
```

File: compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp (hunk 2 above)
Comment (AdvenamTacet) [reply compiler-rt/test/asan/TestCases/contiguous_container_crash.cpp:42]: Update check rules

**Comment by vitalybuka** [comment]
> @vitalybuka do you know how correctly create "does not appear" rule? I added `-NOT-`, but it seems to not work.

I don't like NOT at all. Note, -NOT must be the suffix.

I like --implicit-check-not="PATTERN"

In your case I'd suggest --implicit-check-not="ERROR: AddressSanitizer"
Than ALL instances of this patter must be matches with some CHECK: statement, or test fail.

**Comment by AdvenamTacet** [comment]
Ok, I tested this test in #101349 and it works as intended. I used "implicit-check-not", thanks for pointing me into that, I didn't know it.
I'm going to merge it, unless anyone sees something to fix here.

File: compiler-rt/lib/asan/asan_errors.cpp [hunk 4]
```diff
@@ -328,8 +328,6 @@ void ErrorBadParamsToAnnotateContiguousContainer::Print() {
       "      new_mid : %p\n",
       (void *)beg, (void *)end, (void *)old_mid, (void *)new_mid);
   uptr granularity = ASAN_SHADOW_GRANULARITY;
```
Comment (chapuni) [inline compiler-rt/lib/asan/asan_errors.cpp:330]: This is unused.

File: compiler-rt/lib/asan/asan_errors.cpp [hunk 5]
```diff
@@ -348,8 +346,6 @@ void ErrorBadParamsToAnnotateDoubleEndedContiguousContainer::Print() {
       (void *)old_container_end, (void *)new_container_beg,
       (void *)new_container_end);
   uptr granularity = ASAN_SHADOW_GRANULARITY;
```
Comment (chapuni) [inline compiler-rt/lib/asan/asan_errors.cpp:348]: Ditto.

**Review by chapuni — COMMENTED** [review]
2x2 warnings :)

**Comment by AdvenamTacet** [comment]
@chapuni thank you for fixing them!

**Pull request merged**